src/java/org/apache/cassandra/streaming/messages/IncomingMutationLogStreamMessage.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.replication.MutationTrackingService;
+import org.apache.cassandra.streaming.LogStreamHeader;
+import org.apache.cassandra.streaming.StreamManager;
+import org.apache.cassandra.streaming.StreamReceiveException;
+import org.apache.cassandra.streaming.StreamSession;
+import org.apache.cassandra.streaming.StreamingChannel;
+import org.apache.cassandra.streaming.StreamingDataOutputPlus;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Incoming mutation log stream message for receiving mutation logs during bootstrap.
+ * 
+ * This message handles deserialization and processing of mutation logs received from
+ * nodes during the bootstrap process.
+ */
+public class IncomingMutationLogStreamMessage extends MutationLogStreamMessage
+{
+    private static final Logger logger = LoggerFactory.getLogger(IncomingMutationLogStreamMessage.class);
+
+    public static final StreamMessage.Serializer<IncomingMutationLogStreamMessage> serializer = new IncomingMutationLogStreamMessageSerializer();
+
+    public final StreamSession session;
+
+    public IncomingMutationLogStreamMessage(LogStreamHeader header, StreamSession session)
+    {
+        super(header);
+        this.session = session;
+    }
+
+    @Override
+    public StreamSession getOrCreateAndAttachInboundSession(StreamingChannel channel, int messagingVersion)
+    {
+        session.attachInbound(channel);
+        return session;
+    }
+
+    public static class IncomingMutationLogStreamMessageSerializer implements StreamMessage.Serializer<IncomingMutationLogStreamMessage>
+    {
+        public IncomingMutationLogStreamMessage deserialize(DataInputPlus in, int version) throws IOException
+        {
+            LogStreamHeader header = LogStreamHeader.serializer.deserialize(in, version);
+            StreamSession session = StreamManager.instance.findSession(header.sender, header.planId, header.sessionIndex, header.sendByFollower);
+            if (session == null)
+                throw new IllegalStateException(String.format("unknown stream session: %s - %d", header.planId, header.sessionIndex));
+
+            try
+            {
+                while (in.readBoolean())
+                {
+                    int userVersion = in.readInt();
+                    ByteBuffer buffer = ByteBufferUtil.readWithVIntLength(in);
+                    Mutation mutation = Mutation.serializer.deserialize(buffer, userVersion);
+
+                    if (logger.isTraceEnabled())
+                        logger.trace("Received mutation {}: session={}, keyspace={}, token={}",
+                                     mutation.id(),
+                                     session.planId(),
+                                     mutation.getKeyspaceName(),
+                                     mutation.key().getToken());
+
+                    mutation.apply();
```

Comment:
I was thinking we'd receive all the mutation logs, then do replay to apply all before completing the session, rather than deserializing and applying each at a time